samples/Config.CustomConfigurationProviders.Sample/Microsoft.Extensions.Configuration.ConfigFile/ConfigFileConfigurationProvider.cs

```diff
@@ -0,0 +1,147 @@
+    ///     &lt;/sampleSection&gt;
+    ///   &lt;/configuration&gt;
+    ///
+    /// Code:
+    ///   var configuration = new ConfigurationBuilder().AddConfigFile(&quot;Web.config&quot;).Build();
+    ///   var setting1 = configuration[&quot;sampleSection:setting1&quot;];
+    ///   var allAppSettings = configuration.GetSection(&quot;appSettings&quot;).GetChildren();
+    ///   var myJetConnection = configuration[&quot;connectionStrings:MyJetConn&quot;];
+    /// </example>
+    public class ConfigFileConfigurationProvider : ConfigurationProvider
+    {
+        private readonly ILogger _logger;
+        private readonly string _configuration;
+        private readonly bool _loadFromFile;
+        private readonly bool _isOptional;
+
+        private readonly IEnumerable<IConfigurationParser> _parsers;
+
+        public ConfigFileConfigurationProvider(string configuration, bool loadFromFile, bool optional, params IConfigurationParser[] parsers)
+            : this(configuration, loadFromFile, optional, null, parsers)
+        { }
+
+        public ConfigFileConfigurationProvider(string configuration, bool loadFromFile, bool optional, ILogger logger, params IConfigurationParser[] parsers)
+        {
+            _loadFromFile = loadFromFile;
+            _configuration = configuration;
+            _isOptional = optional;
+            _logger = logger;
+
+            var parsersToUse = new List<IConfigurationParser> {
+                new KeyValueParser(),
+                new KeyValueParser("name", "connectionString")
+            };
+
+            parsersToUse.AddRange(parsers);
+
+            _parsers = parsersToUse.ToArray();
+        }
+
+        public override void Load()
+        {
+            if (_loadFromFile && !_isOptional && !File.Exists(_configuration))
+            {
+                throw new FileNotFoundException("Could not find configuration file to load!", _configuration);
+            }
+
+            XDocument document;
+
+            if (!TryLoadXDocument(out document))
+            {
+                return;
+            }
+
+            Data = new SortedDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            var context = new Stack<string>();
+            var dictionary = new SortedDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var child in document.Root.Elements())
+            {
+                ParseElement(child, ref context, ref dictionary);
+            }
+
+            Data = dictionary;
+        }
+
+        private bool TryLoadXDocument(out XDocument document)
+        {
+            if (!_loadFromFile)
+            {
+                document = XDocument.Parse(_configuration);
+                return true;
+            }
+
+            try
+            {
+                document = XDocument.Load(_configuration);
+                return true;
+            }
+            catch (IOException e)
```

Comment:
I wouldn't catch this exception at all. The problem with catching it here is that if the user doesn't have logging enabled, they'll never know something went wrong. However, this exception really is a fatal exception - the app *must not start* if the configuration data can't be loaded. As such, just let the exception bubble up and let the app die. ASP.NET Core (or whoever the caller is) will deal with the exception and logging on its own.